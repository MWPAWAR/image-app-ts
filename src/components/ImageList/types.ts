import { ImageListTypes, ImageTypes } from '../types'

export interface ImageListProps {
  imageList: ImageListTypes
  toggleDrawer: () => void
}

export interface ImageListItemProps {
  image: ImageTypes
  toggleDrawer: () => void
}
