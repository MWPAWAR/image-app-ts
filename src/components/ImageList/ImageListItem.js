import { useDispatch } from 'react-redux';
import { getMbsTextByBytes } from '../../utils/get-mbs-by-bytes';
import imageListItemStyles from './ImageListItem.module.css';
import { setSelectedImg } from '../../store/imageUiState/imageUiState';

const ImageListItem = ({ image }) => {
  const dispatch = useDispatch();
  const handleImageClick = () => dispatch(setSelectedImg(image.id));

  return (
    <div onClick={handleImageClick} className={imageListItemStyles.imageListItem}>
      <img
        tabIndex={0}
        src={image.url}
        className={imageListItemStyles.imageListItemImg}
        alt={image.filename}
      />
      <div className={imageListItemStyles.imageListItemName}>{image.filename}</div>
      <div className={imageListItemStyles.imageListItemSize}>{getMbsTextByBytes(image.sizeInBytes)}</div>
    </div>
  );
};

export default ImageListItem;
