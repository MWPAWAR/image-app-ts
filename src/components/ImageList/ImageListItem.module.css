.imageListItem {
  cursor: pointer;
  margin: 5px;
  width: 100%;
  text-align: center;
}

.imageListItemImg {
  border-radius: 10px;
  width: 300px;
  height: 150px;
}

.imageListItemImg:focus-within, .imageListItemImg:hover {
  border: 2px solid blue;
}

.imageListItemName {
  font-size: 14px;
  padding: 10px 0 5px 0;
  font-weight: 600;
  overflow: hidden;
  text-overflow: ellipsis;
}

.imageListItemSize {
  font-size: 15px;
  color: #5f6368;
}

@media only screen and (min-width: 481px) {
  .imageListItem {
    width: 200px;
    margin: 15px;
    text-align: left;
  }

  .imageListItemImg {
    border-radius: 10px;
    width: 200px;
    height: 150px;
  }
}
