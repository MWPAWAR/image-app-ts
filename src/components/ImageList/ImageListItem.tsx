import { useAppDispatch } from '../customHooks/hooks'
import { getMbsTextByBytes } from '../../utils/get-mbs-by-bytes'
import imageListItemStyles from './ImageListItem.module.css'
import { setSelectedImg } from '../../store/imageUiState/imageUiState'
import { ImageListItemProps } from './types'

const ImageListItem = ({ image }: ImageListItemProps) => {
  const dispatch = useAppDispatch()
  const handleImageClick = (): void => {
    dispatch(setSelectedImg({ id: image.id }))
  }

  return (
    <div
      data-testid="image-list-item"
      onClick={handleImageClick}
      className={imageListItemStyles.imageListItem}
    >
      <img
        tabIndex={0}
        src={image.url}
        className={imageListItemStyles.imageListItemImg}
        alt={image.filename}
      />
      <div className={imageListItemStyles.imageListItemName}>
        {image.filename}
      </div>
      <div className={imageListItemStyles.imageListItemSize}>
        {getMbsTextByBytes(image.sizeInBytes)}
      </div>
    </div>
  )
}

export default ImageListItem
