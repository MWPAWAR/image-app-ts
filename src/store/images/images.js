import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import getImages from './api';

export const fetchImages = createAsyncThunk('images/fetchImages', async () => {
  const response = await getImages();
  return response;
});

const imagesSlice = createSlice({
  name: 'images',
  initialState: { status: 'idle', images: [] },
  reducers: {
    toggleFavoriteImg(state, action) {
      const index = state.images.findIndex((image) => image.id === action.payload);
      if (typeof index !== 'undefined') {
        state.images[index].favorited = !state.images[index].favorited;
      }
    },
    deleteImg(state, action) {
      const index = state.images.findIndex((image) => image.id === action.payload);
      if (typeof index !== 'undefined') {
        state.images.splice(index, 1);
      }
    }
  },
  extraReducers(builder) {
    builder
      .addCase(fetchImages.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchImages.fulfilled, (state, action) => {
        state.status = 'success';
        state.images = action.payload;
      })
      .addCase(fetchImages.rejected, (state) => {
        state.error = 'error';
      })
  }
});

export const { toggleFavoriteImg, deleteImg } = imagesSlice.actions;
export default imagesSlice.reducer;
