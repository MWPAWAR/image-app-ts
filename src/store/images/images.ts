import { createSlice, createAsyncThunk } from '@reduxjs/toolkit'
import { ImageListTypes } from '../../components/types'
import getImages from './api'
import { API_FETCH_STATUS } from '../../constants'
import {
  ImagesState,
  ImagesExtraReducerStateAction,
  ImagesReducerStateAction,
  ImagesFetchApiResponse,
} from './types'

export const fetchImages = createAsyncThunk<ImagesFetchApiResponse>(
  'images/fetchImages',
  async () => {
    const response: ImageListTypes = await getImages()
    return response
  }
)

const initialState: ImagesState = {
  status: API_FETCH_STATUS.IDLE,
  images: [],
}

const imagesSlice = createSlice({
  name: 'images',
  initialState,
  reducers: {
    toggleFavoriteImg(
      state: ImagesState,
      action: ImagesReducerStateAction
    ): void {
      const index: number | undefined = state.images.findIndex(
        (image) => image.id === action.payload.id
      )
      if (typeof index !== 'undefined') {
        state.images[index].favorited = !state.images[index].favorited
      }
    },
    deleteImg(state: ImagesState, action: ImagesReducerStateAction): void {
      const index: number | undefined = state.images.findIndex(
        (image) => image.id === action.payload.id
      )
      if (typeof index !== 'undefined') {
        state.images.splice(index, 1)
      }
    },
  },
  extraReducers(builder) {
    builder
      .addCase(fetchImages.pending, (state: ImagesState) => {
        state.status = API_FETCH_STATUS.LOADING
      })
      .addCase(
        fetchImages.fulfilled,
        (state: ImagesState, action: ImagesExtraReducerStateAction) => {
          state.status = API_FETCH_STATUS.SUCCESS
          state.images = action.payload
        }
      )
      .addCase(fetchImages.rejected, (state: ImagesState) => {
        state.status = API_FETCH_STATUS.ERROR
      })
  },
})

export const { toggleFavoriteImg, deleteImg } = imagesSlice.actions
export default imagesSlice.reducer
