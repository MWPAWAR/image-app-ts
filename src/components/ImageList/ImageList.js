import ImageListItem from './ImageListItem';
import imageListStyles from './ImageList.module.css';

const ImageList = ({ imageList }) => {
  return (
    <div className={imageListStyles.imageList}>
      {imageList.map(image => <ImageListItem image={image} />)}
    </div>
  );
};

export default ImageList;
