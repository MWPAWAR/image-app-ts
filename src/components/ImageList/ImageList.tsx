import ImageListItem from './ImageListItem'
import imageListStyles from './ImageList.module.css'
import { ImageListProps } from './types'
import { ImageTypes } from '../types'

const ImageList = ({ imageList }: ImageListProps) => (
  <div className={imageListStyles.imageList}>
    {imageList.map((image: ImageTypes) => (
      <ImageListItem key={image.id} image={image} />
    ))}
  </div>
)

export default ImageList
