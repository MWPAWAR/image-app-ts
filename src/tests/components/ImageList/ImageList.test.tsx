import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import ImageList from '../../../components/ImageList/ImageList'
import { images } from '../../mocks/images'

describe('ImageList', () => {
  it('renders ImageList correctly', () => {
    const mockStore = configureStore([])
    const store = mockStore({})

    const ImageListComponent = () => (
      <MemoryRouter>
        <Provider store={store}>
          <ImageList imageList={images} />
        </Provider>
      </MemoryRouter>
    )

    const tree = renderer.create(<ImageListComponent />)
    expect(tree).toMatchSnapshot()
  })
})
