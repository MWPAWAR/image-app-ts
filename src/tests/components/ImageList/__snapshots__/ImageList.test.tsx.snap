// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageList renders ImageList correctly 1`] = `
<div
  className="imageList"
>
  <div
    className="imageListItem"
    data-testid="image-list-item"
    onClick={[Function]}
  >
    <img
      alt="tennessee_female_rubber.jpg"
      className="imageListItemImg"
      src="https://agencyanalytics-api.vercel.app/images/0.jpg"
      tabIndex={0}
    />
    <div
      className="imageListItemName"
    >
      tennessee_female_rubber.jpg
    </div>
    <div
      className="imageListItemSize"
    >
      4.6 MB
    </div>
  </div>
  <div
    className="imageListItem"
    data-testid="image-list-item"
    onClick={[Function]}
  >
    <img
      alt="virginia.jpg"
      className="imageListItemImg"
      src="https://agencyanalytics-api.vercel.app/images/1.jpg"
      tabIndex={0}
    />
    <div
      className="imageListItemName"
    >
      virginia.jpg
    </div>
    <div
      className="imageListItemSize"
    >
      4.1 MB
    </div>
  </div>
</div>
`;
